### ISSUE TITLE
Import one file 

### ISSUE THREAD

**Issue by santiobarco** [joyfullservice/msaccess-vcs-addin#406]
I need update just one file in the aplication, I have a change in one form, what I need to do?

**Comment by joyfullservice** [comment]
@santiobarco - Great question! There are two approaches you can take in the latest beta version. (Version 4.0.13-beta that I just published today)

From the ribbon, you can click the button to `Load Selected`. That will reload that item from the source file.

![image](https://github.com/joyfullservice/msaccess-vcs-addin/assets/7254992/b9f2596a-44a5-428c-8bd1-218730caeafa)

The other option is to perform a `Merge Build`, a newly developed feature which will scan your source files for changes that do not exist in your database, and import any changed source items into the database. (This is typically much faster than rebuilding the entire database from source.)

![image](https://github.com/joyfullservice/msaccess-vcs-addin/assets/7254992/635c6369-bca3-408a-ad31-4c3330a86279)

I hope that helps!

**Comment by santiobarco** [comment]
Thanks for your faster answer, I have an other question, If I want it to compare the existing files and only change the ones that have changes I can do it too, I try to do a merge because I just added some public sub in one of the existing modules and it tells me that I should build from source first

**Comment by joyfullservice** [comment]
Yes, that would be correct. In order to _merge_ from source, you need to do a _full build_ from source first. Once you have done the initial full build, you can export and merge as often as you like.

**Comment by santiobarco** [comment]
Thanks I will try it!!

**Comment by santiobarco** [comment]
This does not make a comparison of a file with another, I mean if a new module was created to a new function, etc.

**Comment by santiobarco** [comment]
Hi @joyfullservice I need to build from source for merge a new changes, but when start the process for build from source the application stays frozen, do you have some advice for make this please

**Comment by joyfullservice** [comment]
Which version of the add-in are you using? Also, are you able to build other databases from source? (For example, if you create a very simple new database file, can you export that to source files, then build it from source?)

**Comment by santiobarco** [comment]
Thanks a lot for the quickly answer I need to solve this now, I use the version 4.0.15-beta and I try to build from source with a simple database and i'ts working with a simple file

**Comment by joyfullservice** [comment]
You could try the build with or without the `Debug VBA Errors` option to see if that helps identify the problem...

![image](https://github.com/joyfullservice/msaccess-vcs-addin/assets/7254992/89cb7f4f-cfcb-4639-ba56-9cf88c7bf415)

When it freezes, are you able to press `ctl`+`break` to stop the running code? Can you tell where the build is struggling to continue? What is the last thing you see on the screen before it gets stuck? These clues may help us pinpoint where the problem is occurring.

**Comment by santiobarco** [comment]
Ok adam I'll try right now!!

**Comment by santiobarco** [comment]
Adam the process is freeze when start the import of the forms

**Comment by joyfullservice** [comment]
Okay, let's try to figure out what is happening with the form import... Here are a couple things you could try:

Using the existing database, try importing the forms one by one using the `Load Selected` button on the toolbar. This might help us see if it is affect all the forms, or just certain ones. To do this, select the form in the navigation pane, then click the `Load Selected` button to reload that individual object from the source file.

A second thing you could try is to go to your simple test database, and use Microsoft Access to import the forms from your big database into your test database. After importing the forms, export the test database to source files, and then try building from source. If it still freezes, then remove all the forms from the test database and try just importing one or two of the most simple forms from the big database.

The goal here is to identify a form that is failing to import, then to figure out what might be causing it to fail. You can also add breakpoints in the VBA code for the form import process (see `clsDbForm`), and step through the code as it performs the import to see if that helps identify exactly where the code is failing.

Let me know what you find out!

**Comment by santiobarco** [comment]
I have more than 100 forms import one by one is not a option, I tried again to build from source and press ctrl + break for see the debug vba Errors but this shortcut is not working.

**Comment by joyfullservice** [comment]
Yes, I can feel your pain with that. The main thing I am trying to identify is if the form import issue is related to ALL of the forms, or just some of them. You probably won't need to import all 100 forms one by one. You might just need to test a few of them to find one that exhibits the issue. From there you can drill down and isolate issues until we can figure out exactly what is causing the import to fail.

If you can find ONE form that fails to import, that you can reproduce with your testing database, it will make troubleshooting the problem a lot easier. My guess is that there will be something we can do to resolve the issue, but we just need to figure out what the problem is first.  😄

**Comment by santiobarco** [comment]
Ok adam I find the form that have issues, how handle this

**Comment by joyfullservice** [comment]
If you want to provide a small sample database, I could test it on my system. Otherwise you will need to set some break points in the VBA add-in code to step through the import process.

Regarding the add-in, try reinstalling it and see if that resolves the block. If not, you might need to figure out how to unblock it again first.

**Comment by santiobarco** [comment]
Adam thanks for the help, when I tried to add-in the VCS, i get this system message is missinig a USysreginfo table, do you know where I need to create this table?

**Comment by joyfullservice** [comment]
That's for a different kind of installation. This add-in doesn't require that table (or admin privileges to install). Just double-click the downloaded add-in file, and click the button to install when it opens.

**Comment by santiobarco** [comment]
Adam but I Made the installation again, but I can't see the VCS in the ribbon is weird

**Comment by joyfullservice** [comment]
> Adam I Made the installation again, but I can't see the VCS in the ribbon is weird

There are two parts to the installation. The Microsoft Access add-in is installed, and available via the ribbon, and via the Add-ins menu. If you can launch the add-in using this menu, you know that the Access add-in is working.

![image](https://github.com/joyfullservice/msaccess-vcs-addin/assets/7254992/e826b71e-ad8a-45ed-bb4b-34921802f9f5)

The second part of the installation is the COM add-in, which is installed a different way. (Access add-ins are literally Access database files with VBA code, while a COM add-in is a compiled DLL, in this case created with twinBASIC.)

To view/manage the COM add-ins, you navigate to the application options:

![image](https://github.com/joyfullservice/msaccess-vcs-addin/assets/7254992/b94087d5-6065-4863-8c30-fa3a220f1bb1)

There you will find the COM add-ins listed, and an option to load/unload or remove.

![image](https://github.com/joyfullservice/msaccess-vcs-addin/assets/7254992/944bfa50-da0d-45af-93a3-9d9b9d043e13)

You mentioned that something was "blocked" at some point when you were testing... I don't know which thing was blocked, or exactly how it was blocked. Normally the installation is very simple. One click and you are done.  😄 You may need to go to the COM add-ins and remove the existing item, then run the (Access add-in) installation again to see if the installer will reinstall it for you.

Again, to install/reinstall, simply double-click `Version Control.accda` and click the `Install` button. That should handle extracting and installing the COM add-in, registering and activating the ribbon, etc...

Hope that helps!

**Comment by santiobarco** [comment]
<img width="632" alt="aa" src="https://github.com/joyfullservice/msaccess-vcs-addin/assets/71194016/7ea7fe3d-5f43-472d-a61f-b9a143408f80">

Adam I made the installation again but look I can't see the VCS in the ribbon

**Comment by joyfullservice** [comment]
What do you see in the COM Add-in dialogs shown in the screenshots above? The COM add-in is what loads the ribbon. If the COM add-in is not loading, you won't see the ribbon.

**Comment by santiobarco** [comment]
<img width="913" alt="image" src="https://github.com/joyfullservice/msaccess-vcs-addin/assets/71194016/2b7d92eb-6b5f-4255-8562-a01d99873609">
Adam I can see the ribbon thanks a lot I'll still investigate the form that have the problem, thanks a lot for them, when I got some news about the form I'll keep in post to you

**Comment by sotalvaro** [comment]
Hi Adam, can you guide me what executes the Build as button I need execute a routine and integrate internally all this flow and will be great see all the internal flow since you hit the Build As button. Thanks for your help man.

**Comment by joyfullservice** [comment]
There are two types of add-ins that work together for this. The COM add-in is what contains the ribbon commands, and is loaded when you open Microsoft Access. This stays loaded all the time. (This is developed in twinBASIC and installed as a compiled DLL.) 

The Access add-in is a Microsoft Access database project that contains the main logic for exporting source files and building a database from source. Unlike the COM add-in which is loaded all the time, this add-in is loaded on demand when it is first used. After the Access add-in is loaded, it stays loaded until the application is closed.

The reason you don't see the `MSAccessVCS` project in the screenshot above is because it has not yet been loaded in this instance of Access. When you click one of the VCS ribbon commands, the Access add-in will be automatically loaded, and you will see it listed in the VBA projects list. The internal Access wizards work the same way. Once you run a wizard, that add-in stays loaded in Microsoft Access until the application is closed.

To find the code that is executed when you click a button on the ribbon, take a look at the `clsVersionControl` class module. There you can find the `BuildAs()` function, and see exactly what the code is doing.

You mentioned needing to execute a routine and internally integrate this workflow. I am not sure that I completely understand what you are wanting to do, but here are a couple things that may be helpful:

- This add-in includes some "hooks" to execute code in your own database at various points in the export and build process. Using [these hooks](https://github.com/joyfullservice/msaccess-vcs-addin/wiki/Options#build), you can automate a lot of internal processes such as cleanup and build verification.
- If you are wanting to externally automate the build process itself, this is something that is not yet fully developed at this point.

> I need to overwrite always with the same name of the original version, just need to build the new app with the same original name and overwrite, for don't create more copies inside the folder.

A normal build does exactly that, replacing the existing database with a freshly built version. The "copies" in the folder are backups of the database before each build. If you really don't want the backup files, you can add some code to delete the backups and trigger it to run after the build process.

Hope that helps!

**Comment by sotalvaro** [comment]
Hi Adam Thanks for the quickly answer, here is the thing that I need to do, I need to run the build as process automatically, because I set up a pipeline in azure dev ops and the release open the access file, when the access file is open I need to validate the user that is open the file (The user of the pipeline of azure dev ops) after that I need to, run the build as process always in the same path and overwrite the file with the same name.

**Comment by joyfullservice** [comment]
As of the latest commit, you should have everything you need to do this, but you will obviously need to work through the specifics in your particular build process. To launch the add-in and initiate actions, please see `ExampleLoadAddInAndRunExport()` in `modAPI`. The specific commands you will want to run are:

```VBA
MSAccessVCS.SetInteractionMode 1   ' Set to silent
MSAccessVCS.HandleRibbonCommand "btnBuild"
```

This will run the same action as clicking the "Build From Source" button on the ribbon toolbar. (To build the current database from source.)

I would highly recommend using the standard build operation, not the Build As for automated builds. You can check out the project to a build folder in your deployment pipeline, and then after building from source, you can test/verify your file, then push it out to whatever distribution system you have set up. You may also want to add some automated checks to review the build log file to check for any errors.

If you have further questions/issues with the build automation, please feel free to open a new issue for that. I think we have pretty well resolved the original issue of importing one file.  😄 

I am planning to roll out a new dev release today (4.0.21), so that should include all of the latest commits needed to perform an automated build.

**Comment by sotalvaro** [comment]
Adam Thanks for the answer. If I need to execute the same routine that you post in there I can see the in the same method? but i need to change just this MSAccessVCS.Exportsource for what word man. 
<img width="757" alt="image" src="https://github.com/joyfullservice/msaccess-vcs-addin/assets/138826677/94171de6-a475-4a08-96ee-a3b67894ec27">
And i'll create the new issue hehe
